Make specification creation and mixing easier, remove boilerplate from specifications

graphicore opened the pull request:
WIP! Announced here: https://github.com/googlefonts/fontbakery/pull/1757#issuecomment-376911081


This introduces a new and completely optioninal style of defining specifications. Instead of using `@register_check / specification.register_check` and `@register_condition / specification.register_condition` these can be automatically detected from the module namespace and registered in the spec.

Further, a spec can be created automatically from a module. *Currently* (more will follow) fully functional is [`specifcations/cmap`](https://github.com/graphicore/fontbakery/blob/420f4b2105b245edb2dd98c2bb1b951c2fa9f90c/Lib/fontbakery/specifications/cmap.py) in this PR, which does itself neither define a `specification` nor use any `@register_` decorators. It's the simplest  possible definition of a spec, yet:

```sh 
$ fontbakery check-specification  fontbakery.specifications.cmap ../testfonts/*
```

is just working.

There is a tool to mix other modules into the current spec `specification.register_modules(*specification_modules)`

to make something like this possible:

```py
from . import (general, cmap, head, os2, post, name, hhea, dsig, hmtx, gpos,
               gdef, kern, glyf, prep, fvar, shared_conditions)

# specification = ...
specification_modules= (general, cmap, head, os2, post, name, hhea, dsig, hmtx, gpos,
               gdef, kern, glyf, prep, fvar, shared_conditions)
specification.register_modules(*specification_modules)
```
Of course, explicit registration of checks, conditions is still possible just as before.

To make this possible I also added `expected_values` to specification, we can't use names in dependency injection anymore if they are not registered in the spec namespace. `expected_values`
is not the actual values but a placeholder-name with description and optionally default value and validator.

This has some big advantages:
  * it's stricly documented what values a spec needs to run
  * there's a proper way to validate user input!
  * the documentation of a spec can output a proper values description

also see the detailed message in 9a1f68b2e2


Lastly, there are some tests available now, to warn if a spec is incomplete:
* self-test: test_dependencies Raises SetupError if spec uses any names that are not declared in its namespace.
* optional, spec.test_expected_checks: if auto_register is used, this helps a spec maintainer to be aware of changes in the spec

graphicore commented:
Next I will make all specification working, especially the `googlefonts` one is interesting, because it takes in all other specs and because it has some quirkiness I want to get rid of.

One annoyance is for example `whitelist_librebarcode` because it's very google fonts specific, yet used all over the place.

madig commented:
```
from . import (general, cmap, head, os2, post, name, hhea, dsig, hmtx, gpos,
               gdef, kern, glyf, prep, fvar, shared_conditions)
# specification = ...
specification.register_modules(*specification_modules)
```
Do you mean
```
from . import (general, cmap, head, os2, post, name, hhea, dsig, hmtx, gpos,
               gdef, kern, glyf, prep, fvar, shared_conditions)
specification_modules= (general, cmap, head, os2, post, name, hhea, dsig, hmtx, gpos,
               gdef, kern, glyf, prep, fvar, shared_conditions)
specification.register_modules(*specification_modules)
```
?

graphicore commented:
> Do you mean

Yes, totally! Sorry it was very early.

madig commented:
I like the pattern :) It would be even nicer to not have to repeat the modules. Maybe register_modules can look up the names as strings using importlib, looking in `.` by default but also accepting e.g. `"fontbakery.specification.my_special_table"`?

graphicore commented:
> I like the pattern :)

cool.

>  It would be even nicer to not have to repeat the modules.

Yes. I'm not sure about using file system location though. We could have a special marker in the module namespace, like the name "`specification`" or if, not present the name "`spec_factory`" and that being a function (callable) the latter could then be used to create the right Spec type. 

`spec.auto_register` could check all python `<module ,,,>` types for these markers. I'd add more control to `auto_register` so we can tell it to look e.g. only for modules.

I'm also suggesting `spec_factory` to tackle:

https://github.com/graphicore/fontbakery/blob/420f4b2105b245edb2dd98c2bb1b951c2fa9f90c/Lib/fontbakery/commands/check_spec.py#L12-L18

madig commented:
Or maybe look in `fontbakery.specifications` by default when a module name has no other qualifier.

madig commented:
So, the end goal would be to able to do something like `fontbakery check-specification cmap fonts*ttf` as well as `fontbakery check-specification googlefonts fonts*ttf` where the composition can happen under the hood, right?

graphicore commented:
Yes. But it needs to be controllable what goes into a spec, automatic composition is optional.

graphicore commented:
@madig any thoughts/ideas about this:

https://github.com/graphicore/fontbakery/blob/420f4b2105b245edb2dd98c2bb1b951c2fa9f90c/Lib/fontbakery/specifications/cmap.py#L8

```py
from .shared_conditions import ttFont

# this silences flake8 (F401 'ttFont' imported but unused and
# F811 redefinition of unused 'ttFont' from line 6), because we're not
# really using ttFont, but it's needed for auto_register external to this module
(ttFont, )
```

I want namespace items (conditions, expected_values, iterargs, ...) to be explicit, so that there's no room for interpretation if there are colliding names. Hence a spec module, to be self contained must import/define its conditons. This is correct python but flake8 goes crazy :-/

graphicore commented:
well, after "using" `ttFont` in 

```py
(ttFont, )
```
flake8 is happy, but I think its rather annoying to pull this.

graphicore commented:
In the case of `ttFont` it is probably best to register it directly in [`fonts_spec`](https://github.com/graphicore/fontbakery/blob/420f4b2105b245edb2dd98c2bb1b951c2fa9f90c/Lib/fontbakery/fonts_spec.py#L62) since there we mention it first:

```py
 , derived_iterables={'ttFonts': ('ttFont', True)}
```

Still, other namespace items will have the same problem.

madig commented:
Can you change the line to `from .shared_conditions import ttFont  # noqa: F401` and make flake8 not complain? Seems cleaner.

graphicore commented:
No, that way it still complains about F811 multiple times, for each `@conditon` that uses `ttFont` as argument (which is just wrong and should be fixed in flake8)

graphicore commented:
With 2e1ef6ad2a7 all specs work now with `fontbakery check-spec fontbakery.specification.*`

@madig: Note the changes in `ufo_sources.py` (added `ExpectedValue` definition for `fonts`)

Besides the flake8 issue, we now get a good idea which conditions are required by each spec. I also added a `googlefonts_shared_conditions` module, since otherwise we have a circular import problem.

madig commented:
(I recommend you use yapf for formatting new files and do not format by hand, e.g. by removing the double space between imports and the first function definition. Formatting by hand is a waste of time and yapf ensures that everyone is on the same page regarding formatting. Doing it by default on new files does not lead to the commit churn you'd see if you did it on existing files)

graphicore commented:
(isn't that a bit nit-picking "the double space between imports and the first function definition"?)

madig commented:
(It is a detail, yes. One that allows you to brainlessly run yapf without ending up with commit churn :)

graphicore commented:
(you don't have to reformat it and cause commit churn, just leave it as it is and go on with your life ;)

madig commented:
(Well, you did and wasted some seconds of your life :P I write shoddily formatted code and let yapf clean up after me, lets me focus on more important things)

graphicore commented:
(You got to practice your zen and focus on the details, LOL)

madig commented on Lib/fontbakery/specifications/name.py:
```diff
@@ -7,6 +7,11 @@
+from .shared_conditions import seems_monospaced, monospace_stats
+from .googlefonts_shared_conditions import whitelist_librebarcode
+# flake8 F401, F811:
+(seems_monospaced, monospace_stats, whitelist_librebarcode)
```
Hm... should the monospace things rellay be Google-specific?

madig commented:
Okay, the changes look like an improvement to me :) merge when? :D

madig commented:
Hm. Maybe FontsSpec should be changed to accept both files and directories. Maybe specifications should define their own setup_argparse argument checking functions?

graphicore commented:
> merge when

I still want to do more things. However, I just hit a strange heisenbug (probably something with my terminal emulator) and that's setting me back. And I'm away in an hour and running out of time today :-/

graphicore commented:
@felipesanches please take a look at 1d96c77 as you are the main maintainer of the googlefonts spec.

This now uses the auto_register feature, so I was able to remove all `@register_*` decorators.
To make sure nothing got lost, I also used the self test `specification.test_expected_checks` with the `exclusive=True` option. That means that the available checks must be the same as in `expected_check_ids` no more, no less. `exclusive=False` would allow more checks than in `expected_check_ids` specified, so the spec could grow automatically, if another spec module would add a new check.
The `auto_register` + `exclusive=true` option had however already a positive effect, I discovered that from the previous breakup in 81c50fb the test `com_google_fonts_check_010` was not registered in `googlefonts` anymore.

We'll see how to best use this is day to day maintenance. But for now I feel that `exclusive=true`gives the best overview and control over what is going on in the spec, with the burden of having to add new tests to the `expected_check_ids` list.

One big question if we wan't to use the `auto_register` style OR if we want to go with explicit registering by hand (and remove this commit from the PR).

In the meantime, I'll go on and improve the tooling for `auto_register` to also detect spec modules, as described in https://github.com/googlefonts/fontbakery/pull/1761#issuecomment-377203674 ff. and probably to take a custom filter function.

felipesanches commented:
@graphicore Thanks for taking care of adding the `@disable` decorator for inhibiting the registering of `com.google.fonts/check/082` !

felipesanches commented:
I tend to dislike the extra burden of listing all new checks... I'd prefer to keep it simple.

graphicore commented:
This is about control vs. simplicity. I'd personally chose control, but I can't say which is better. One perspective is that we now removed explicit import + registration and added explicit "expected_check_ids" so it is  a zero sum.

We don't have to use `test_expected_checks` but I think it's good to have and I can imagine that there is a demand for it, e.g. if a third party maintains a custom spec and needs to be informed when the QA tool changes.

graphicore commented:
@madig  about https://github.com/googlefonts/fontbakery/pull/1761#issuecomment-377304660

> Maybe FontsSpec should be changed to accept both files and directories.

I'm not sure, glob does the job, or? 

> Maybe specifications should define their own setup_argparse argument checking functions? Important for directories is that the path is normalized, as the user can pass in some.ufo/ on the commandline just from tabbing and .endswith won't trigger.

I'm not sure if I understand. `UFOSpec` already defines its own `setup_argparse` different from `FontsSpec`

graphicore commented:
Here's my rest todo-list for this PR:

* When auto_registering checks, we should be able blacklist certain check id's or cherry pick (maybe also namespace items). Filter functions?
* some mechanism to mark `blacklisted_for_this_check` I'm not happy having `whitelist_librebarcode` spread through many general specs.
* maybe something about `googlefonts_shared_conditions.style` as well, as it is used in the general section
* move fontforge_check_results from `googlefonts_shared_conditions` into `general`, as it is used only there
* with the thre items above `googlefonts_shared_conditions` would be eliminated here as it defines only `whitelist_librebarcode`, `fontforge_check_results` and `style`)
* Maybe: things like `whitelist_camelcased_familyname` could use CAMELCASED_FAMILYNAME as ExpectedValue. We should be able to use the current value for google fonts as default,  but create this way the possibility to configure it using `values`. There are no comandline args at this point, just a way to do it.
* remove `test_expected_checks` from `googlefonts`?

madig commented:
> glob does the job, or? I don't see the need to accept directories vs mydir/*

You need to accept directories for UFOs :)

> I'm not sure if I understand. UFOSpec already defines its own setup_argparse different from FontsSpec

Hm, on a second though, keeping UFOSpec with its own setup_argparse might be okay. It's the only situation where you're looking at directories I guess.

madig commented:
> maybe something about googlefonts_shared_conditions.style as well, as it is used in the general section

Don't forget about custom, foundry-specific style names that exist due to policy, so defining acceptable styles yourself would be very nice to have :)

graphicore commented:
> foundry-specific style names

yeah, that's a tricky one ;-)

graphicore commented:
The main thin that happened is that we can now do:

```py
@condition(force=True)
def ...
```

To override and existing namespace item with the condition. That way, a spec can define a generic or placeholder condition and another spec (the googlefonts spec) can define it's private overrides. The `style` and (partly) `whitelist_librebarcode conditon` issues were solved with this. And other foundries can write their own `style` conditions.

The other big thing is that with `specification.set_check_filter(check_filter)` a filter function can be registered to blacklist check executions under certain conditions. 

also `Spec.auto_register` can now take a filter function, to blacklist certain items. This is not used at the moment, but it is needed for more complete control over the `auto_register` process.


The last to items on my todo-list  are still open, but not so urgent, I will think about it tomorrow.

graphicore commented:
@madig inspired by https://github.com/googlefonts/fontbakery/pull/1761#issuecomment-377203674

> Maybe register_modules can look up the names as strings using importlib

we can now do:

```py
spec_imports = (
    ('.', ('general', 'cmap', 'head', 'os2', 'post', 'name',
       'hhea', 'dsig', 'hmtx', 'gpos', 'gdef', 'kern', 'glyf',
       'prep', 'fvar', 'shared_conditions')
    ),
)
```

instead of:

```py
from . import (general, cmap, head, os2, post, name, hhea, dsig, hmtx, gpos,
               gdef, kern, glyf, prep, fvar, shared_conditions) # NOQA

(general, cmap, head, os2, post, name, hhea, dsig, hmtx, gpos,
               gdef, kern, glyf, prep, fvar, shared_conditions)

 ```

and 

```py
spec_imports = [
    ('.shared_conditions', ('is_variable_font'
            , 'regular_wght_coord', 'regular_wdth_coord', 'regular_slnt_coord'
            , 'regular_ital_coord', 'regular_opsz_coord', 'bold_wght_coord'))
]


```

instead of:

```py
from .shared_conditions import (is_variable_font
            , regular_wght_coord, regular_wdth_coord, regular_slnt_coord
            , regular_ital_coord, regular_opsz_coord, bold_wght_coord)
# flake8 F401, F811:
(is_variable_font
            , regular_wght_coord, regular_wdth_coord, regular_slnt_coord
            , regular_ital_coord, regular_opsz_coord, bold_wght_coord)
```

graphicore commented:
> * Maybe: things like whitelist_camelcased_familyname could use CAMELCASED_FAMILYNAME as ExpectedValue. We should be able to use the current value for google fonts as default, but create this way the possibility to configure it using values. There are no comandline args at this point, just a way to do it.

I won't do this in this PR. We have a bunch of tools to handle overrides now. If someone wants to make a condition configurable it is possible in different ways.

> * remove test_expected_checks from googlefonts?

@felipesanches this is your call. I'm confident that I did not lose any checks while refactoring :-D

graphicore commented:
Apart from that, if there are no requests for changes, I'd like to suggest to merge this.

davelab6 commented:
I think @felipesanches might be recovering from his trip back home from Boston :)

madig commented:
@graphicore one thing: please check your code with `pip install --user future && futurize all/sourcefiles/you/touched/*.py` and see if it changes something. Use www.python-future.org/futurize.html as a guide. This will make the Py3 transition easier!

graphicore commented:
@madig LOL, the main purpose of that tool seems to be wrapping each iterated item into a `list(...)`.

madig commented:
> yeah, that's a tricky one ;-)

Hm. how about changing the mapping dict from a spec module? So if I go and make a daltonmaagfonts spec, I could set fontbakery.constants.STYLE_NAMES to what I want. If the spec is not run, stuff isn't changed.

madig commented:
Well, Py2 and 3 differ in their iterator approach :D If the code happens to work on Py3, maybe you can do all the other things it finds? Like change the lambda syntax to what Py3 expects.

madig commented on Lib/fontbakery/specifications/googlefonts.py:
```diff
@@ -5424,6 +2972,183 @@ def com_google_fonts_check_173(ttFont):
+
+# this is from the output of
+# $ fontbakery check-specification  fontbakery.specifications.googlefonts -L
+expected_check_ids = [
```
Just a cosmetic thing, but maybe it looks nicer if you put the expected tests near the top of the file? The check if everything is registered can stay at the bottom.

graphicore replied:
good point I would do so if we keep it; but @felipesanches was not very fond of this anyways. Let's see what he says.

graphicore commented:
I didn't touch any of these cases, but there's a way now:

You would define a `Condition` or an `ExpectedValue(..., default=...` called `STYLE_NAMES`. Then use the dependency injection (`condition` or `check` argument names).

A default could be implemented and then you spec would override it with `force=true` OR if you use an `ExpectedValue` you can set it via the CheckRunner and it's `values` argument (we could add a cli argument to check-specifcation to load values from a json/yaml/whatever file and set them to check runner or you create your own check-daltonmaag command)

madig commented:
Sounds meta 😁 

By the way, what's the test coverage? If you rebase to master, run `tox` and then `coverage html`, you'll get a nice HTML report in htmlcov/ that shows you which lines are tested and which aren't.

graphicore commented:
You want me to post that `coverage html` report omewhere?

```
Name                                                 Stmts   Miss Branch BrPart  Cover
--------------------------------------------------------------------------------------
Lib/fontbakery/__init__.py                               0      0      0      0   100%
Lib/fontbakery/callable.py                             122     10     32      4    90%
Lib/fontbakery/checkrunner.py                          833    250    389     86    66%
Lib/fontbakery/commands/__init__.py                      0      0      0      0   100%
Lib/fontbakery/commands/check_googlefonts.py            11     11      0      0     0%
Lib/fontbakery/commands/check_spec.py                   86     86     26      0     0%
Lib/fontbakery/commands/check_ufo_sources.py            11     11      0      0     0%
Lib/fontbakery/constants.py                             73      0      0      0   100%
Lib/fontbakery/fonts_public_pb2.py                      18      0      0      0   100%
Lib/fontbakery/fonts_spec.py                            28     16      4      0    38%
Lib/fontbakery/glyphdata.py                              1      0      0      0   100%
Lib/fontbakery/message.py                                9      2      0      0    78%
Lib/fontbakery/reporters/__init__.py                    61     61     12      0     0%
Lib/fontbakery/reporters/serialize.py                   83     83     44      0     0%
Lib/fontbakery/reporters/terminal.py                   291    291    124      0     0%
Lib/fontbakery/specifications/__init__.py                0      0      0      0   100%
Lib/fontbakery/specifications/cmap.py                   54     31     42      2    34%
Lib/fontbakery/specifications/dsig.py                    8      0      2      0   100%
Lib/fontbakery/specifications/fvar.py                   29      0     12      0   100%
Lib/fontbakery/specifications/gdef.py                   17      4      8      2    68%
Lib/fontbakery/specifications/general.py               245    124    128      6    47%
Lib/fontbakery/specifications/glyf.py                   28     22     14      0    14%
Lib/fontbakery/specifications/googlefonts.py          1190    612    572     30    46%
Lib/fontbakery/specifications/googlefonts_test.py     1320     22    252     11    98%
Lib/fontbakery/specifications/gpos.py                   46     11     38      4    68%
Lib/fontbakery/specifications/head.py                  104     73     44      0    28%
Lib/fontbakery/specifications/hhea.py                   41     27     22      0    29%
Lib/fontbakery/specifications/hmtx.py                   19     12      4      0    30%
Lib/fontbakery/specifications/kern.py                    8      0      2      0   100%
Lib/fontbakery/specifications/name.py                  156     13    102      7    90%
Lib/fontbakery/specifications/os2.py                    83     31     48      0    61%
Lib/fontbakery/specifications/post.py                   26      0     12      0   100%
Lib/fontbakery/specifications/prep.py                   11      6      4      0    33%
Lib/fontbakery/specifications/shared_conditions.py      84      4     34      4    93%
Lib/fontbakery/specifications/ufo_sources.py            73     21     26      1    74%
Lib/fontbakery/specifications/ufo_sources_test.py       73      0      0      0   100%
Lib/fontbakery/utils.py                                 85     19     40      3    79%
--------------------------------------------------------------------------------------
TOTAL                                                 5327   1853   2037    160    62%
__________________________________________ summary ___________________________________________
```

graphicore commented:
By the way, I did some of the `futurize` suggestions, though, I don't think all of them are needed. I skipped most of the `list(...)` wraps.

graphicore commented:
> Sounds meta :grin:

If course whole `@check`/`@condition` DI-stuff is very meta, otherwise we wouldn't know how to run a check for each font when all information we have is `def mycheck(ttFont)` ...

anthrotype commented:
> You want me to post that coverage html report omewhere?

you can use codecov.io like we do in https://codecov.io/gh/fonttools/fonttools and others related repos

madig commented:
> You want me to post that coverage html report somewhere?

No, I just thought that you had a look at how much your code is covered by the tests and maybe write some if important paths aren't covered ;)

graphicore commented:
I think we only have good tests for the specifications. Coverage is a bit misleading here. The infrastructure is not unit tested yet.

felipesanches commented:
I'm continuing work on this at #1770
